### DISCUSSION TITLE
Error in Android after installing react-native-twitter-signin/twitter-signin library with rnfirebase

### DISCUSSION THREAD

**Discussion by anctsys** [invertase/react-native-firebase#6115] [General]
After install the library 
npm i @react-native-twitter-signin/twitter-signin
Follow the procedure  https://rnfirebase.io/auth/social-auth

Installation broken

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:checkDebugAarMetadata'.
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Could not find com.twitter.sdk.android:twitter-core:3.3.0.
     Searched in the following locations:
       - file:/node_modules/react-native/android/com/twitter/sdk/android/twitter-core/3.3.0/twitter-core-3.3.0.pom
       - file:/node_modules/jsc-android/dist/com/twitter/sdk/android/twitter-core/3.3.0/twitter-core-3.3.0.pom
       - https://repo.maven.apache.org/maven2/com/twitter/sdk/android/twitter-core/3.3.0/twitter-core-3.3.0.pom
       - https://dl.google.com/dl/android/maven2/com/twitter/sdk/android/twitter-core/3.3.0/twitter-core-3.3.0.pom
       - https://www.jitpack.io/com/twitter/sdk/android/twitter-core/3.3.0/twitter-core-3.3.0.pom
     Required by:
         project :app > project :react-native-twitter-signin_twitter-signin
```

Do you have a solution to use social authentification twitter with rnfirebase

**Comment by mikehardy** [comment]
Your title says google, your error says twitter. I will assume the title is wrong and change it (correct me if wrong)

I have not personally used Twitter auth but I seem to recall that Twitter was pulling their native SDKs and no longer supported them?

In fact they are archived however they still appear to be available. Here's what I turned up for artifact information on the 3.3.0 release:

https://mvnrepository.com/artifact/com.twitter.sdk.android/twitter-core?repo=springio-libs-release

So in the springiolibs-release repository you may still find this release, the URL for that repo is visible here, if you check this page:
https://mvnrepository.com/repos/springio-libs-release

If you open `android/build.gradle` and add a new repository entry for that URL you should be able to access the artifact and get a successful build.

If that works for you (I hope it does!), then you should post a PR to the readme on the react-native-twitter-signin repo detailing the new requirement to edit android/build.gradle and add a 3rd party repo to get the android lib

**Reply by kirillpisarev** [reply]
> Your title says google, your error says twitter. I will assume the title is wrong and change it (correct me if wrong)
> 
> I have not personally used Twitter auth but I seem to recall that Twitter was pulling their native SDKs and no longer supported them?
> 
> In fact they are archived however they still appear to be available. Here's what I turned up for artifact information on the 3.3.0 release:
> 
> https://mvnrepository.com/artifact/com.twitter.sdk.android/twitter-core?repo=springio-libs-release
> 
> So in the springiolibs-release repository you may still find this release, the URL for that repo is visible here, if you check this page: https://mvnrepository.com/repos/springio-libs-release
> 
> If you open `android/build.gradle` and add a new repository entry for that URL you should be able to access the artifact and get a successful build.
> 
> If that works for you (I hope it does!), then you should post a PR to the readme on the react-native-twitter-signin repo detailing the new requirement to edit android/build.gradle and add a 3rd party repo to get the android lib

Looks like they restricted the anonymous access to `/libs-release` - https://spring.io/blog/2020/10/29/notice-of-permissions-changes-to-repo-spring-io-fall-and-winter-2020 . As i understood only the Spring Team members now allowed to use it this way.

So using that approach now reports `Could not HEAD 'https://repo.spring.io/libs-release/com/twitter/sdk/android/twitter-core/3.3.0/twitter-core-3.3.0.pom'. Received status code 401 from server: Unauthorized`

Seems like the only way to use twitter sign in in RN now is to build the SDK from [sources](https://github.com/GoldenOwlAsia/react-native-twitter-signin) and then use `mavenLocal`

**Reply by mikehardy** [reply]
I believe that's inaccurate, I think twitter does not do custom auth anymore by that SDK but supports oath, so the modern pathway is here, via oath2 to twitter using this module

https://github.com/fullstackreact/react-native-oauth/blob/3063886c162758278a0f54fe423ba74730bd1d92/lib/authProviders.js#L24-L36

Much better than trying to patch up explicitly disavowed code with escalating workarounds

**Reply by mikehardy** [reply]
Unfortunately - I think my comment is technically correct but also incomplete: I don't think we have support here (yet) for generic oauth provider in order to connect with the oauth library

It was started but needs a dev (commits by Ilja) https://github.com/invertase/react-native-firebase/commits/%40ilja/twitter-auth

https://firebase.google.com/docs/reference/android/com/google/firebase/auth/OAuthProvider
https://firebase.google.com/docs/reference/swift/firebaseauth/api/reference/Classes/OAuthProvider

**Reply by kirillpisarev** [reply]
You got ahead of my question - "why do we need 3rd party libs to handle client side token receiving, when firebase already handles it". So for anyone wondering, there are [startActivityForLinkWithProvider](https://firebase.google.com/docs/auth/android/twitter-login?hl=en) on android and [getCredentialWithUIDelegate -> signInWithCredential](https://firebase.google.com/docs/auth/ios/twitter-login?hl=en) on iOS to do so.

I personally tried to implement the fast and dirty patch that leverages those methods instead of using the unsupported twitter lib, but for some reason i didn't manage to retrieve access token from `AuthResult` on android (probably because of the old firebase version) so and we ended up removing twitter oauth from our app, but that is another story. I believe if the twitter oauth is must have, that approach is pretty viable.

**Reply by mikehardy** [reply]
Yep - just generic oauth is something that would be a truly fantastic addition to the library and I'm personally sad we don't have it yet, I would be happy to collaborate with anyone getting it going. Unfortunately I've got higher priority "full library" work like getting the v9 firebase-js-sdk syntax going and handling the firebase-ios-sdk v9 breaking changes before I can take on module-specific items, so it could be quite a while if no one helps

**Reply by kmsbernard** [reply]
I've also ran into this problem, and just start to implement native SDK methods which @kirillpisarev mentioned above. I'll open a PR as work progress.

**Reply by mikehardy** [reply]
Oh that's exciting to hear! thank you thank you! I'm travelling at the moment on vacation so won't be a great collaborator for around a week but please don't mistake that for lack of enthusiasm on this feature
